core-services/src/main/java/org/zalando/nakadi/service/publishing/AvroEventPublisher.java

```diff
@@ -0,0 +1,156 @@
+                serializeAndSendToKafka(etName, metadata, event);
+
+            } catch (final InterruptedException e) {
+                Thread.currentThread().interrupt();
+                LOG.error("Failed to wait for timeline switch", e);
+            } catch (final TimeoutException e) {
+                LOG.error("Failed to wait for timeline switch", e);
+            } finally {
+                try {
+                    if (publishingCloser != null) {
+                        publishingCloser.close();
+                    }
+                } catch (final IOException e) {
+                    LOG.error("Exception occurred when releasing usage of event-type", e);
+                }
+            }
+
+        } catch (final Exception e) {
+            LOG.error("Error occurred when submitting KPI event for publishing", e);
+        }
+    }
+
+    private void serializeAndSendToKafka(final String eventTypeName,
+                                         final GenericRecord eventMetadata,
+                                         final GenericRecord event) {
+        final EventType eventType = eventTypeCache.getEventType(eventTypeName);
+        final Timeline activeTimeline = timelineService.getActiveTimeline(eventType);
+        final String topic = activeTimeline.getTopic();
+
+        final Span publishingSpan = TracingService.buildNewSpan("publishing_to_kafka")
+                .withTag(Tags.MESSAGE_BUS_DESTINATION.getKey(), topic)
+                .withTag("event_type", eventTypeName)
+                .withTag("type", "avro")
+                .start();
+
+        try (Closeable ignored = TracingService.activateSpan(publishingSpan)) {
+            final ByteArrayOutputStream eventOutputStream = new ByteArrayOutputStream();
+            final GenericDatumWriter metadataWriter = new GenericDatumWriter(eventMetadata.getSchema());
+            metadataWriter.setSchema(event.getSchema());
+            metadataWriter.write(event, EncoderFactory.get()
+                    .directBinaryEncoder(eventOutputStream, null));
+            final byte[] metadata = eventOutputStream.toByteArray();
+            eventOutputStream.reset();
+
+            final GenericDatumWriter eventWriter = new GenericDatumWriter(event.getSchema());
+            eventWriter.write(event, EncoderFactory.get()
+                    .directBinaryEncoder(eventOutputStream, null));
+            final byte[] payload = eventOutputStream.toByteArray();
+            final byte[] envelope = Envelope.serialize(new Envelope(
+                    AvroSchema.METADATA_VERSION,
+                    metadata.length,
+                    metadata,
+                    payload.length,
+                    payload)
+            );
+
+            timelineService.getTopicRepository(eventType)
+                    .syncPostEvent(new NakadiRecord(
+                            eventTypeName,
+                            topic,
+                            null,
```

Comment:
Do I get it right, that using `null` for partition results in random partitioning?

`If a valid partition number is specified that partition will be used when sending the record. If no partition is specified but a key is present a partition will be chosen using a hash of the key. If neither key nor partition is present a partition will be assigned in a round-robin fashion.`
https://kafka.apache.org/23/javadoc/org/apache/kafka/clients/producer/ProducerRecord.html#ProducerRecord-java.lang.String-java.lang.Integer-K-V-

Reply:
yes, for the moment I chose that approach.